Fix instruments not filtered by method in Worksheet Template edit view

Lunga001 opened the pull request:
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/

## Current behavior before PR
On the worksheet template on edit mode, selecting a method does not filter the instrument that belong to selected method
## Desired behavior after PR is merged
 Method selection should filter the instrument by the selected method
 
--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

Lunga001 committed:
Remove SelectionWidget on RestrictToMethod field to avoid last option selection behavior on the dropdown

Lunga001 committed:
Show instruments available for the selected method

ramonski commented:
Hi @Lunga001, what is the state of this PR?

Lunga001 commented:
Updated @ramonski the PR is ready. Sorry about the delay, just that I wasn't sure removing selection widget was the best way to go about it.

Lunga001 committed:
Merge branch '2.x' into wttemplate-method-select

Lunga001 committed:
Changelog

xispa commented on src/bika/lims/content/worksheettemplate.py:
```diff
@@ -100,16 +99,6 @@
-                "should save the change first."
-            ),
-        ),
     ),
```
Removing this widget makes the filtering of instruments when a method is selected to not work at all. The reason is that the javascript in charge of updating the list of methods expects the control to be a `SelectionWidget` here: https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/browser/static/js/bika.lims.worksheettemplate.js#L11

xispa commented on src/bika/lims/browser/worksheettemplate.py:
```diff
@@ -31,16 +29,20 @@ class AJAXGetWorksheetTemplateInstruments(BrowserView):
+
+        try:
+            instruments = api.get_object(method_uid).getInstruments()
+        except api.APIError:
```
Better remove this try/except. System might fail silently otherwise.

xispa reviewed (changes requested):
Thanks @Lunga001, but it does not work, see comments. 

Note this `AJAXGetWorksheetTemplateInstruments` class and the javascript are old-school. You might consider to use the "EditForm" introduced with https://github.com/senaite/senaite.core/pull/1719 to achieve this instrument-filtering functionality in a more elegant way, and with extrapoints for the removal of old code.

Lunga001 committed:
Using adapter edit form for worksheettemplate

Lunga001 committed:
Revert "Remove SelectionWidget on RestrictToMethod field to avoid last option selection behavior on the dropdown"

This reverts commit e1f4023b11d4e5314e186d85513804b6bb9d4fb4.

Lunga001 commented:
> Thanks @Lunga001, but it does not work, see comments.
> 
> Note this `AJAXGetWorksheetTemplateInstruments` class and the javascript are old-school. You might consider to use the "EditForm" introduced with #1719 to achieve this instrument-filtering functionality in a more elegant way, and with extrapoints for the removal of old code.

Gave EditForm a shot

Lunga001 committed:
Remove unused files

xispa committed:
Cleanup getInstruments and getRawInstruments getters from worksheet template

xispa commented on src/senaite/core/browser/form/adapters/configure.zcml:
```diff
@@ -32,6 +32,20 @@
+           zope.publisher.interfaces.browser.IBrowserRequest"
+      factory=".method.ValidateInstrumentsAdapterBase"/>
+  -->
+
```
Please remove this instead if commenting if is not used

xispa commented on src/senaite/core/browser/form/adapters/worksheettemplate.py:
```diff
@@ -0,0 +1,76 @@
+            # Available Methods
+            empty = [{"title": _("No Instrument"), "value": [""]}]
+            # Get selected method
+            method = api.get_object_by_uid(value[0])
```
Note that `value[0]` can be empty. You get `APIError: get_object_by_uid requires UID as first argument; got  instead` in such case. Remember you can check if something is an UID with `api.is_uid(someting)`

xispa commented on src/senaite/core/browser/form/adapters/worksheettemplate.py:
```diff
@@ -0,0 +1,76 @@
+            empty = [{"title": _("No Instrument"), "value": [""]}]
+            # Get selected method
+            method = api.get_object_by_uid(value[0])
+            instruments = self.get_available_instruments_for(method)
```
You can omit this `get_available_instruments_for` and rely on method's `getInstruments` directly here, cause in accordance with my previous comment, at this point you should have a valid method already

xispa commented on src/senaite/core/browser/form/adapters/worksheettemplate.py:
```diff
@@ -0,0 +1,76 @@
+                self.add_update_field("Instruments:list", {
+                    "options": []})
+                self.add_update_field("Instruments_options", {
+                    "options": i_opts})
```
this should be empty instead of `i_opts`, right?

xispa reviewed (changes requested):
@Lunga001 please pull, note I've pushed a commit for cleanup here https://github.com/senaite/senaite.core/pull/1887/commits/3312335b1fd072b02fc375e8cadc16a33ec55125

I also left some comments

Lunga001 committed:
Use api.is_ui to check for empty methods
Omit get_available_instruments_for and rely on method's getInstruments
Remove commented code

xispa committed:
Imports cleanup

xispa committed:
Remove WorksheetTemplateEdit from loader.js

xispa committed:
Changelog

xispa committed:
Simplify

xispa committed:
Fix method selected in selection list when no method assigned

xispa committed:
Cleanup
